src/avram/model.cr

```diff
@@ -184,6 +184,79 @@ class Avram::Model
     {% COLUMNS << {name: type_declaration.var, type: data_type, nilable: nilable.id, autogenerated: autogenerated} %}
   end
 
+  macro polymorphic_belongs_to(type_declaration, autogenerated = false)
+
+    {% nilable = false %}
+    {% for type in type_declaration.type.types %}
+      {% if type.class_name == "Path" %}
+        {% nilable = true if type.names[0] == "Nil" %}
+      {% elsif type.class_name == "Union" %}
+        {% nilable = true if type.types.any? { |item| item.names[0] == "Nil" } %}
+      {% end %}
+    {% end %}
+
+    {% COLUMNS << {name: type_declaration.var + "_id", type: Int64, nilable: nilable.id, autogenerated: autogenerated} %}
+    {% COLUMNS << {name: type_declaration.var + "_type", type: String, nilable: nilable.id, autogenerated: autogenerated} %}
+
+    macro finished
+      class SaveOperation
+        before_save :validates_polymorphic_association
+
+        private def validates_polymorphic_association
+          previous_def
+
+          {% if nilable %}
+          validate_inclusion_of {{ type_declaration.var }}_type, in: {{type_declaration.type.types.map(&.stringify).push("").push(nil)}}
+          {% else %}
+          validate_required {{ type_declaration.var }}_type
+          validate_required {{ type_declaration.var }}_id
+          validate_inclusion_of {{ type_declaration.var }}_type, in: {{type_declaration.type.types.map(&.stringify)}}
+          {% end %}
+        end
+      end
+    end
+
+    def {{type_declaration.var}} : {{type_declaration.type}}{% if nilable %}?{% end %}
+    {% if nilable %}
```

Comment:
I'm horrible at reading macros, so I'm going to assume this is all good 😂

Reply:
😂 😂 😂 it's hard to read, needs to assume it's all good reading the tests that covers it hahaha